File: lib/recourse/schedule.ex
defmodule Recourse.Schedule do
  alias Recourse.Registration
  alias Recourse.Repo
  alias Recourse.Section
  use Aruspex.Constraint
  import Recourse.Digraph
  import Ecto.Query
  import __MODULE__.Constraint

  def build(%{"course_ids" => course_ids, "settings" => settings}) do
    {:ok, pid} = Aruspex.start_link

    # required for Registration.load
    {:ok, _} = Recourse.Scraper.start

    course_ids
    |> sections_query
    |> Repo.all
    |> Registration.load
    |> init_variables(pid)
    |> init_constraints(settings, pid)

    Aruspex.find_solution(pid)
    |> Dict.values
  end

  @spec components(Section.t) :: [[Section.t]]
  def components sections do
    g = :digraph.new

    values =
      for section <- sections,
          mt <- section.meeting_times,
          day <- mt.days do
        %{section | meeting_times: [%{mt | days: [day]}]}
        |> add_vertex(g).()
      end

    for x <- values, y <- values, x != y do
      if no_conflict([x, y]) > 0 do
        add_edge g, x, y
      end
    end

    :digraph_utils.components g
  end

  defp init_variables(sections, pid) do
    sections
    |> Enum.group_by(& {&1.course_id, &1.schedule_type}) # %{ int: [Section] }
    |> Enum.map(fn {grouping, sections} ->
      Aruspex.variable(pid, grouping, sections)
      grouping
    end)
  end

  defp init_constraints(variables, settings, pid) do
    for x <- variables, y <- variables, x > y do
      Aruspex.post pid, constraint(
        variables: [x, y],
        function: &no_conflict/1)
    end

    for v <- variables do
      Aruspex.post pid, constraint(
        variables: [v],
        function: time_preference(settings))
    end

    for v <- variables do
      Aruspex.post pid, constraint(
        variables: [v],
        function: &open_seats/1)
    end

    :ok
  end

  defp sections_query course_ids do
    from(s in Section,
      select: s,
      join: c in assoc(s, :course),
      join: mt in assoc(s, :meeting_times),
      where: c.id in ^course_ids,
      where: not(is_nil(mt.start_time) and is_nil(mt.end_time)),
      preload: [meeting_times: [:section], course: :term],
      order_by: c.id)
  end
end

File: apps/pysolver/pysolver/schedule.py
import itertools


class MeetingTime:
    def __init__(self, date_start, date_end, start_time, days, end_time):
        self.date_end = date_end
        self.date_start = date_start
        self.days = days
        self.end_time = end_time
        self.start_time = start_time


class Section:
    def __init__(self, name, schedule_type, course_id, id, meeting_times):
        self.name = name
        self.schedule_type = schedule_type
        self.course_id = course_id
        self.id = id
        self.meeting_times = [MeetingTime(**meeting_time) for
                              meeting_time in meeting_times]


class Group:
    def __init__(self, course_id, schedule_type, sections):
        self.course_id = course_id
        self.schedule_type = schedule_type
        self.sections = [Section(**section) for section in sections]


class Schedule:
    def solve(sections):
        groups = Schedule.group_by_schedule_type(sections)
        return groups

    def key(section):
        return (section['course_id'],
                section['schedule_type'])

    def group_by_schedule_type(sections):
        sections = sorted(sections, key=Schedule.key)
        grouped_sections = itertools.groupby(sections, key=Schedule.key)
        result = list()

        for (course_id, schedule_type), sections in grouped_sections:
            result.append(Group(course_id, schedule_type, list(sections)))

        return result
